scripts/update_appup.escript

```diff
@@ -1,75 +1,313 @@
+added to each patch release preceding the current release. If an entry
+for a module already exists, this module is ignored. The existing
+actions are kept.
+
+Note: The defaults are set up for emqx, but they can be tuned to
+support other repos too.
+
+Usage:
+
+   update_appup.escript [--check] [--repo URL] [--remote NAME] [--skip-build] [--make-commad SCRIPT] [--release-dir DIR] <current_release_tag>
+
+Options:
+
+  --check         Don't update the appup files, just check that they are complete
+  --repo          Upsteam git repo URL
+  --remote        Get upstream repo URL from the specified git remote
+  --skip-build    Don't rebuild the releases. May produce wrong results
+  --make-command  A command used to assemble the release
+  --release-dir   Release directory
+".
+
+default_options() ->
+    #{ check        => false
+     , clone_url    => find_upstream_repo("origin")
+     , make_command => "make emqx-rel"
+     , beams_dir    => "_build/emqx/rel/emqx/lib/"
+     }.
+
+main(Args) ->
+    #{current_release := CurrentRelease} = Options = parse_args(Args, default_options()),
+    case find_pred_tag(CurrentRelease) of
+        {ok, Baseline} ->
+            main(Options, Baseline);
+        undefined ->
+            log("No appup update is needed for this release, nothing to be done~n", []),
+            ok
     end.
 
-erl_basename(App, Name) ->
-    case filename:basename(Name, ".erl") of
-        Name   -> false;
-        Module -> {true, {list_to_atom(App), list_to_atom(Module)}}
+parse_args([CurrentRelease = [A|_]], State) when A =/= $- ->
+    State#{current_release => CurrentRelease};
+parse_args(["--check"|Rest], State) ->
+    parse_args(Rest, State#{check => true});
+parse_args(["--skip-build"|Rest], State) ->
+    parse_args(Rest, State#{make_command => "true"});
+parse_args(["--repo", Repo|Rest], State) ->
+    parse_args(Rest, State#{clone_url => Repo});
+parse_args(["--remote", Remote|Rest], State) ->
+    parse_args(Rest, State#{clone_url => find_upstream_repo(Remote)});
+parse_args(["--make-command", Command|Rest], State) ->
+    parse_args(Rest, State#{make_command => Command});
+parse_args(["--release-dir", Dir|Rest], State) ->
+    parse_args(Rest, State#{beams_dir => Dir});
+parse_args(_, _) ->
+    fail(usage()).
+
+main(Options = #{check := Check}, Baseline) ->
+    {CurrDir, PredDir} = prepare(Baseline, Options),
+    CurrBeams = hashsums(find_beams(CurrDir)),
+    PredBeams = hashsums(find_beams(PredDir)),
+    Upgrade = diff_releases(CurrBeams, PredBeams),
+    Downgrade = diff_releases(PredBeams, CurrBeams),
+    Apps = maps:keys(Upgrade),
+    lists:foreach( fun(App) ->
+                           #{App := AppUpgrade} = Upgrade,
+                           #{App := AppDowngrade} = Downgrade,
+                           process_app(Baseline, Check, App, AppUpgrade, AppDowngrade)
+                   end
+                 , Apps
+                 ).
```

Comment:
Maybe at the end of main, I think it is nice to print some warning/reminder that generated app src should be manually checked since loading some modules could cause process get restarted, and some internal states need to be upgraded, order of module loading should be reviewed too.

Reply:
also NIF loading, I guess